log/logger_impl.go

```diff
@@ -0,0 +1,132 @@
+package log
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+)
+
+func NewDefaultLogger() Logger {
+	return NewLogger(log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr)))
+}
+
+func NewNopLogger() Logger {
+	return NewLogger(log.NewNopLogger())
+}
+
+func NewJSONLogger() Logger {
+	return NewLogger(log.NewJSONLogger(log.NewSyncWriter(os.Stderr)))
+}
+
+func NewBufferLogger() (*strings.Builder, Logger) {
+	buffer := strings.Builder{}
+	writer := log.NewLogfmtLogger(log.NewSyncWriter(&buffer))
+	log := NewLogger(writer)
+	return &buffer, log
+}
+
+func NewLogger(writer log.Logger) Logger {
+	l := &logger{
+		writer:  writer,
+		ctx:     map[string]string{},
+		lastErr: nil,
+	}
+
+	// Default logs to be info until changed
+	return l.Info()
+}
+
+type logger struct {
+	writer  log.Logger
+	ctx     map[string]string
+	lastErr error
+}
+
+func (l *logger) Set(key, value string) Logger {
+	return l.With(Fields{
```

Comment:
I don't think we need to call `With` here.

We can do
```go
l.ctx[key] = value
return l
```

Reply:
Mutating it without creating a new instance of it can cause issues with thread-safe-ness. Using `l.With` is a way of re-using the correct pattern for adding values.